### PR TITLE
Pyproxy mixins

### PR DESCRIPTION
**Pull request by hoodmane**
This PR improves several things about `PyProxy`:

1. Reduce stray properties to the absolute minimum. As much as possible attribute access should go to the Python object.
2. Extra methods defined on Proxies of particular types of Python objects should overridden by properties on the python object (as discussed in #1175)
3. Share prototypes between PyProxies of the same type to reduce memory usage.
4. Organization, explanations, more consistent behavior.

Prior to this PR, fields called "name", "length", or "prototype" all showed weird behavior ("prototype" would be shadowed entirely, "name" and "length" show inconsistent results with the "in" keyword). Likewise "caller" and "arguments" were in all cases shadowed. Now the "name" and "length" inconsistencies are fixed, "prototype" is no longer shadowed (and setting and deleting the "prototype" field works as expected) but "prototype" shows some minor (and unremoveable) inconsistency with the "in" keyword. "caller" and "arguments" are now only shadowed if the `PyProxy` wraps a callable Python object, in other cases they behave normally.

I also added a lot of comments and organized and simplified the logic for testing which abstract object protocols the Python object supports.

This is on top of #1175.

### CODE REVIEW

File: src/core/pyproxy.c [hunk 1]
```diff
@@ -13,6 +13,136 @@ _Py_IDENTIFIER(add_done_callback);
...
+//
+// Object protocol wrappers
+//
+// This section defines wrappers for Python Object protocol API calls that we
+// are planning to offer on the PyProxy. Much of this could be written in
+// Javascript instead. Some reasons to do it in C:
+//  1. Some CPython APIs are actually secretly macros which cannot be used from
+//     Javascript.
+//  2. The code is a bit more concise in C.
+//  3. It may be preferable to minimize the number of times we cross between
```
Comment (phorward) [inline src/core/pyproxy.c:132]: IMHO we should avoid talking about "not being sure" about something in the code. I think the solution as it is now has its right to exist, therefore it is fine.

**Review by phorward — COMMENTED** [review]
Hello @hoodmane ,
thank for this pull request, it looks very impressive. From my first view I don't  see any problems, the tests look nice on the mixins and the documentation is very well. The only "little" thing which bothers me are the comments about when you're not sure about something. I think these kind of thoughts must not be part of the source code, but that's only my opinion.

Separating the Javascript-parts from pyproxy.c into a new pyproxy.js could also be done within this pull request, but IMHO having all related C-JS-stuff in one file is okay.

**Comment by phorward** [comment]
Addresses #1049, #1006, #693, I'll do a test soon on my original issue :+1:

**Comment by hoodmane** [comment]
@phorward Thanks for the review.

I don't think this PR addresses the GC issues. I'll probably do that after this is merged.

**Comment by phorward** [comment]
> I don't think this PR addresses the GC issues. I'll probably do that after this is merged.

Yes, I found this already out, and I'm very excited about that change.

File: src/core/pyproxy.c (hunk 1 above)
Comment (phorward) [reply src/core/pyproxy.c:132]: @hoodmane what do you think about this?

Comment (hoodmane) [reply src/core/pyproxy.c:132]: I'm happy to change the comment. Should I just remove "I'm not sure"?

**Comment by phorward** [comment]
Thank you @hoodmane that is ok for me. Let's wait for CI, and can you please merge the current master into your branch?

File: src/core/pyproxy.c (hunk 1 above)
Comment (hoodmane) [reply src/core/pyproxy.c:132]: In general I think it is fine to have comments stating things like "I'm not sure". For reference, CPython 3.8.2 source has 27 comments saying variations of "I'm not sure", the emscripten source has 11.

**Comment by hoodmane** [comment]
So what's the status on this? Is this okay to merge @phorward?

**Pull request merged**